File: examples/go/cart/logic/cart_state.go
package logic

type CartItem struct {
	ProductID      string
	Name           string
	Quantity       int32
	UnitPriceCents int32
}

type CartState struct {
	CustomerID    string
	Items         map[string]*CartItem // productID -> item
	SubtotalCents int32
	CouponCode    string
	DiscountCents int32
	Status        string // "active", "checked_out"
}

func (s *CartState) Exists() bool {
	return s.CustomerID != ""
}

func (s *CartState) IsActive() bool {
	return s.Status == "active"
}

func (s *CartState) CalculateSubtotal() int32 {
	var subtotal int32
	for _, item := range s.Items {
		subtotal += item.Quantity * item.UnitPriceCents
	}
	return subtotal
}

func EmptyState() *CartState {
	return &CartState{
		Items: make(map[string]*CartItem),
	}
}

File: examples/csharp/Cart/src/CartState.cs
using Examples;

namespace Angzarr.Examples.Cart;

public record CartState(
    string CustomerId,
    List<LineItem> Items,
    int SubtotalCents,
    string CouponCode,
    int DiscountCents,
    string Status)
{
    public static CartState Empty => new("", new List<LineItem>(), 0, "", 0, "");

    public bool Exists => !string.IsNullOrEmpty(CustomerId);
    public bool IsActive => Status == "active";
    public bool IsCheckedOut => Status == "checked_out";
}
